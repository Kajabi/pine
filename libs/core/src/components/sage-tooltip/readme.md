# my-component



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description                                                        | Type                                                                                                                                                                 | Default     |
| ----------- | ------------ | ------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `content`   | `content`    | Content for the tooltip. If HTML is required, use the content slot | `""`                                                                                                                                                                 | `undefined` |
| `hasArrow`  | `has-arrow`  | Determines whether or not the tooltip have an arrow                | `boolean`                                                                                                                                                            | `undefined` |
| `isVisible` | `is-visible` | Determines whether or not the tooltip is visible                   | `boolean`                                                                                                                                                            | `false`     |
| `placement` | `placement`  | Determines the preferred position of the tooltip                   | `"bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'top'`     |


## Events

| Event      | Description                       | Type               |
| ---------- | --------------------------------- | ------------------ |
| `sageHide` | Emitted after a tooltip is closed | `CustomEvent<any>` |
| `sageShow` | Emitted after a tooltip is shown  | `CustomEvent<any>` |


## Methods

### `hideTooltip() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `showTooltip() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Slots

| Slot        | Description                  |
| ----------- | ---------------------------- |
|             | The tooltip's target element |
| `"content"` | Content inside the tooltip   |


## Shadow Parts

| Part        | Description |
| ----------- | ----------- |
| `"arrow"`   |             |
| `"content"` |             |


----------------------------------------------


