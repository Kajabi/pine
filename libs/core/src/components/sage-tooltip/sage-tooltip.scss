:host {
  --background-color: #000;
  --box-shadow: 0 8px 14px rgba(0,0,0,0.16), 0 0 4px rgba(0,0,0,0.08);
  --color: #fff;
  --width: 236px;
  --arrow-size: 6px;
  --arrow-offset: 14px;

  --overlay-border-radius: 10px;
  --overlay-font-size: 14px;
  --overlay-line-height: 20px;
  --overlay-padding: 8px 12px;

  --arrow-pointing-down: var(--arrow-size) var(--arrow-size) 0;
  --arrow-pointing-to-the-right: var(--arrow-size) 0 var(--arrow-size) var(--arrow-size);
  --arrow-pointing-to-the-left: var(--arrow-size) var(--arrow-size) var(--arrow-size) 0;
  --arrow-pointing-up: 0 var(--arrow-size) var(--arrow-size);

  display: inline-block;
  position: relative;

  ::slotted([slot="content"]) {
    white-space: normal;
    width: var(--width);
  }
}

.sage-tooltip__content {
  background-color: var(--background-color);
  border-radius: var(--overlay-border-radius);
  box-shadow: var(--box-shadow);
  color: var(--color);
  font-size: var(--overlay-font-size);
  line-height: var(--overlay-line-height);
  opacity: 0;
  padding: var(--overlay-padding);
  position: absolute;
  visibility: hidden;
  width: max-content;
  z-index: 1;

  :host(.sage-tooltip--has-html-content) & {
    width: auto;
  }

  &::after {
    border-color: transparent;
    border-right-color: transparent;
    border-style: solid;
    border-width: var(--arrow-pointing-to-the-left);
    content: '';
    height: 0;
    position: absolute;
    width: 0;

    .sage-tooltip--right & {
      border-inline-end-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-left);
      left: calc(var(--arrow-size) * -1);
      top: 50%;
      transform: translateY(-50%);
    }

    .sage-tooltip--right-end & {
      border-inline-end-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-left);
      bottom: var(--arrow-offset);
      left: calc(var(--arrow-size) * -1);
      top: initial;
    }

    .sage-tooltip--right-start & {
      border-inline-end-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-left);
      left: calc(var(--arrow-size) * -1);
      top: var(--arrow-offset);
    }

    .sage-tooltip--top & {
      border-block-start-color: var(--background-color);
      border-width: var(--arrow-pointing-down);
      bottom: calc(var(--arrow-size) * -1);
      left: 50%;
      top: initial;
      transform: translateX(-50%);
    }

    .sage-tooltip--top-start & {
      border-block-start-color: var(--background-color);
      border-width: var(--arrow-pointing-down);
      bottom: calc(var(--arrow-size) * -1);
      left: var(--arrow-offset);
      top: initial;
    }

    .sage-tooltip--top-end & {
      border-block-start-color: var(--background-color);
      border-width: var(--arrow-pointing-down);
      bottom: calc(var(--arrow-size) * -1);
      left: initial;
      right: var(--arrow-offset);
      top: initial;
    }

    .sage-tooltip--left & {
      border-inline-start-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-right);
      left: initial;
      right: calc(var(--arrow-size) * -1);
      top: 50%;
      transform: translateY(-50%);
    }

    .sage-tooltip--left-end & {
      border-inline-start-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-right);
      bottom: var(--arrow-offset);
      left: initial;
      right: calc(var(--arrow-size) * -1);
      top: initial;
    }

    .sage-tooltip--left-start & {
      border-inline-start-color: var(--background-color);
      border-width: var(--arrow-pointing-to-the-right);
      left: initial;
      right: calc(var(--arrow-size) * -1);
      top: var(--arrow-offset);
    }

    .sage-tooltip--bottom & {
      border-block-end-color: var(--background-color);
      border-width: var(--arrow-pointing-up);
      left: 50%;
      top: calc(var(--arrow-size) * -1);
      transform: translateX(-50%);
    }

    .sage-tooltip--bottom-end & {
      border-block-end-color: var(--background-color);
      border-width: var(--arrow-pointing-up);
      left: initial;
      right: var(--arrow-offset);
      top: calc(var(--arrow-size) * -1);
    }

    .sage-tooltip--bottom-start & {
      border-block-end-color: var(--background-color);
      border-width: var(--arrow-pointing-up);
      left: var(--arrow-offset);
      top: calc(var(--arrow-size) * -1);
    }
  }
}
