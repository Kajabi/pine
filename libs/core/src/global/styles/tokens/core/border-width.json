{
  "border-width": {
    "none": {
      "value": "0"
    },
    "thin": {
      "value": "1px"
    },
    "thick": {
      "value": "2px"
    }
  }

}