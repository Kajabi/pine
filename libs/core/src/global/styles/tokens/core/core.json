{
  "border-radius": {
    "0": {
      "value": "0",
      "type": "borderRadius"
    },
    "50": {
      "value": "4px",
      "type": "borderRadius"
    },
    "75": {
      "value": "6px",
      "type": "borderRadius"
    },
    "100": {
      "value": "8px",
      "type": "borderRadius"
    },
    "125": {
      "value": "10px",
      "type": "borderRadius"
    },
    "150": {
      "value": "12px",
      "type": "borderRadius"
    },
    "175": {
      "value": "14px",
      "type": "borderRadius"
    },
    "200": {
      "value": "16px",
      "type": "borderRadius"
    },
    "225": {
      "value": "18px",
      "type": "borderRadius"
    },
    "250": {
      "value": "20px",
      "type": "borderRadius"
    },
    "275": {
      "value": "22px",
      "type": "borderRadius"
    },
    "300": {
      "value": "24px",
      "type": "borderRadius"
    },
    "round": {
      "value": "9999px",
      "type": "borderRadius"
    }
  },
  "border-width": {
    "none": {
      "value": "0",
      "type": "borderWidth"
    },
    "thin": {
      "value": "1px",
      "type": "borderWidth"
    },
    "thick": {
      "value": "2px",
      "type": "borderWidth"
    }
  },
  "box-shadow": {
    "xs": {
      "value": [
        {
          "x": "0",
          "y": "1px",
          "blur": "2px",
          "spread": "0",
          "color": "rgba(0,0,0,0.06)",
          "type": "dropShadow"
        },
        {
          "x": "0",
          "y": "1px",
          "blur": "3px",
          "spread": "0",
          "color": "rgba(0,0,0,0.1)",
          "type": "dropShadow"
        }
      ],
      "type": "boxShadow"
    },
    "sm": {
      "value": [
        {
          "x": "0",
          "y": "2px",
          "blur": "4px",
          "spread": "0",
          "color": "rgba(0,0,0,0.12)",
          "type": "dropShadow"
        },
        {
          "x": "0",
          "y": "0",
          "blur": "2px",
          "spread": "0",
          "color": "rgba(0,0,0,0.08)",
          "type": "dropShadow"
        }
      ],
      "type": "boxShadow"
    },
    "md": {
      "value": [
        {
          "x": "0",
          "y": "8px",
          "blur": "14px",
          "spread": "0",
          "color": "rgba(0,0,0,0.16)",
          "type": "dropShadow"
        },
        {
          "x": "0",
          "y": "0",
          "blur": "4px",
          "spread": "0",
          "color": "rgba(0,0,0,0.08)",
          "type": "dropShadow"
        }
      ],
      "type": "boxShadow"
    },
    "lg": {
      "value": {
        "x": "0",
        "y": "8px",
        "blur": "40px",
        "spread": "0",
        "color": "rgba(0,0,0,0.24)",
        "type": "dropShadow"
      },
      "type": "boxShadow"
    }
  },
  "color": {
    "white": {
      "value": "#FFF",
      "type": "color"
    },
    "black": {
      "value": "#000",
      "type": "color"
    },
    "grey": {
      "50": {
        "value": "#FCFCFC",
        "type": "color"
      },
      "100": {
        "value": "#F8F8F8",
        "type": "color"
      },
      "150": {
        "value": "#F0F0F0",
        "type": "color"
      },
      "200": {
        "value": "#E4E4E4",
        "type": "color"
      },
      "300": {
        "value": "#D2D1D1",
        "type": "color"
      },
      "400": {
        "value": "#BBBAB9",
        "type": "color"
      },
      "500": {
        "value": "#9B9A98",
        "type": "color"
      },
      "600": {
        "value": "#828180",
        "type": "color"
      },
      "700": {
        "value": "#6C6A69",
        "type": "color"
      },
      "800": {
        "value": "#4D4D4C",
        "type": "color"
      },
      "900": {
        "value": "#343332",
        "type": "color"
      },
      "950": {
        "value": "#1A1A19",
        "type": "color"
      }
    },
    "blue": {
      "50": {
        "value": "#FAFCFF",
        "type": "color"
      },
      "100": {
        "value": "#EFF6FF",
        "type": "color"
      },
      "150": {
        "value": "#DBE9FE",
        "type": "color"
      },
      "200": {
        "value": "#BFDBFE",
        "type": "color"
      },
      "300": {
        "value": "#92C5FD",
        "type": "color"
      },
      "400": {
        "value": "#60A5FA",
        "type": "color"
      },
      "500": {
        "value": "#3C82F6",
        "type": "color"
      },
      "600": {
        "value": "#2463EB",
        "type": "color"
      },
      "700": {
        "value": "#1C4ED8",
        "type": "color"
      },
      "800": {
        "value": "#1D40AE",
        "type": "color"
      },
      "900": {
        "value": "#1F3A8A",
        "type": "color"
      },
      "950": {
        "value": "#172554",
        "type": "color"
      }
    },
    "green": {
      "50": {
        "value": "#FBFEFC",
        "type": "color"
      },
      "100": {
        "value": "#EDFCF2",
        "type": "color"
      },
      "150": {
        "value": "#D3F8DF",
        "type": "color"
      },
      "200": {
        "value": "#AAF0C4",
        "type": "color"
      },
      "300": {
        "value": "#73E2A3",
        "type": "color"
      },
      "400": {
        "value": "#3DCB7F",
        "type": "color"
      },
      "500": {
        "value": "#17B365",
        "type": "color"
      },
      "600": {
        "value": "#079250",
        "type": "color"
      },
      "700": {
        "value": "#097443",
        "type": "color"
      },
      "800": {
        "value": "#085C37",
        "type": "color"
      },
      "900": {
        "value": "#094C2F",
        "type": "color"
      },
      "950": {
        "value": "#052E1C",
        "type": "color"
      }
    },
    "red": {
      "50": {
        "value": "#FFFAFA",
        "type": "color"
      },
      "100": {
        "value": "#FEF2F2",
        "type": "color"
      },
      "150": {
        "value": "#FEE2E1",
        "type": "color"
      },
      "200": {
        "value": "#FECACA",
        "type": "color"
      },
      "300": {
        "value": "#FDA5A5",
        "type": "color"
      },
      "400": {
        "value": "#F87171",
        "type": "color"
      },
      "500": {
        "value": "#EF4444",
        "type": "color"
      },
      "600": {
        "value": "#DC2625",
        "type": "color"
      },
      "700": {
        "value": "#B91C1B",
        "type": "color"
      },
      "800": {
        "value": "#991B1B",
        "type": "color"
      },
      "900": {
        "value": "#7F1C1D",
        "type": "color"
      },
      "950": {
        "value": "#572627",
        "type": "color"
      }
    },
    "yellow": {
      "50": {
        "value": "#FFFEFA",
        "type": "color"
      },
      "100": {
        "value": "#FFFBEB",
        "type": "color"
      },
      "150": {
        "value": "#FFF3C6",
        "type": "color"
      },
      "200": {
        "value": "#FEE589",
        "type": "color"
      },
      "300": {
        "value": "#FED04B",
        "type": "color"
      },
      "400": {
        "value": "#FDBB21",
        "type": "color"
      },
      "500": {
        "value": "#F79A0B",
        "type": "color"
      },
      "600": {
        "value": "#DB7303",
        "type": "color"
      },
      "700": {
        "value": "#B64F06",
        "type": "color"
      },
      "800": {
        "value": "#943D0E",
        "type": "color"
      },
      "900": {
        "value": "#79330E",
        "type": "color"
      },
      "950": {
        "value": "#451902",
        "type": "color"
      }
    },
    "mercury": {
      "50": {
        "value": "#FFFCFA",
        "type": "color"
      },
      "100": {
        "value": "#FFF3ED",
        "type": "color"
      },
      "150": {
        "value": "#FFE3D4",
        "type": "color"
      },
      "200": {
        "value": "#FFC3A8",
        "type": "color"
      },
      "300": {
        "value": "#FF9970",
        "type": "color"
      },
      "400": {
        "value": "#FF6337",
        "type": "color"
      },
      "500": {
        "value": "#FF3E15",
        "type": "color"
      },
      "600": {
        "value": "#F11F06",
        "type": "color"
      },
      "700": {
        "value": "#C81307",
        "type": "color"
      },
      "800": {
        "value": "#9E110E",
        "type": "color"
      },
      "900": {
        "value": "#7F120F",
        "type": "color"
      },
      "950": {
        "value": "#450506",
        "type": "color"
      }
    },
    "purple": {
      "50": {
        "value": "#FAFBFF",
        "type": "color"
      },
      "100": {
        "value": "#EDF1FF",
        "type": "color"
      },
      "150": {
        "value": "#E0E4FF",
        "type": "color"
      },
      "200": {
        "value": "#C7CDFE",
        "type": "color"
      },
      "300": {
        "value": "#A3ACFD",
        "type": "color"
      },
      "400": {
        "value": "#8081F9",
        "type": "color"
      },
      "500": {
        "value": "#6A62F2",
        "type": "color"
      },
      "600": {
        "value": "#533BE5",
        "type": "color"
      },
      "700": {
        "value": "#4F37CB",
        "type": "color"
      },
      "800": {
        "value": "#402FA4",
        "type": "color"
      },
      "900": {
        "value": "#372D82",
        "type": "color"
      },
      "950": {
        "value": "#221B4B",
        "type": "color"
      }
    }
  },
  "font-family": {
    "greet": {
      "value": "\"GreetStandard\"",
      "type": "fontFamilies"
    }
  },
  "font-size": {
    "57": {
      "value": "8px",
      "type": "fontSizes"
    },
    "71": {
      "value": "10px",
      "type": "fontSizes"
    },
    "85": {
      "value": "12px",
      "type": "fontSizes"
    },
    "100": {
      "value": "14px",
      "type": "fontSizes"
    },
    "116": {
      "value": "16px",
      "type": "fontSizes"
    },
    "128": {
      "value": "18px",
      "type": "fontSizes"
    },
    "142": {
      "value": "20px",
      "type": "fontSizes"
    },
    "157": {
      "value": "22px",
      "type": "fontSizes"
    },
    "171": {
      "value": "24px",
      "type": "fontSizes"
    },
    "185": {
      "value": "26px",
      "type": "fontSizes"
    },
    "200": {
      "value": "28px",
      "type": "fontSizes"
    },
    "214": {
      "value": "30px",
      "type": "fontSizes"
    },
    "228": {
      "value": "32px",
      "type": "fontSizes"
    },
    "242": {
      "value": "34px",
      "type": "fontSizes"
    },
    "257": {
      "value": "36px",
      "type": "fontSizes"
    },
    "271": {
      "value": "38px",
      "type": "fontSizes"
    },
    "285": {
      "value": "40px",
      "type": "fontSizes"
    }
  },
  "font-weight": {
    "thin": {
      "value": "125",
      "type": "fontWeights"
    },
    "extra-light": {
      "value": "225",
      "type": "fontWeights"
    },
    "light": {
      "value": "325",
      "type": "fontWeights"
    },
    "normal": {
      "value": "425",
      "type": "fontWeights"
    },
    "medium": {
      "value": "525",
      "type": "fontWeights"
    },
    "semi-bold": {
      "value": "625",
      "type": "fontWeights"
    },
    "bold": {
      "value": "725",
      "type": "fontWeights"
    },
    "extra-bold": {
      "value": "825",
      "type": "fontWeights"
    },
    "heavy": {
      "value": "925",
      "type": "fontWeights"
    }
  },
  "line-height": {
    "25": {
      "value": "0.625",
      "type": "lineHeights"
    },
    "50": {
      "value": "0.78125",
      "type": "lineHeights"
    },
    "75": {
      "value": "0.9375",
      "type": "lineHeights"
    },
    "100": {
      "value": "1",
      "type": "lineHeights"
    },
    "125": {
      "value": "1.25",
      "type": "lineHeights"
    },
    "142": {
      "value": "1.425",
      "type": "lineHeights"
    },
    "150": {
      "value": "1.5",
      "type": "lineHeights"
    },
    "175": {
      "value": "1.75",
      "type": "lineHeights"
    },
    "200": {
      "value": "2",
      "type": "lineHeights"
    },
    "225": {
      "value": "2.25",
      "type": "lineHeights"
    },
    "250": {
      "value": "2.5",
      "type": "lineHeights"
    },
    "275": {
      "value": "2.75",
      "type": "lineHeights"
    },
    "300": {
      "value": "3",
      "type": "lineHeights"
    }
  },
  "spacing": {
    "0": {
      "value": "0",
      "type": "spacing"
    },
    "50": {
      "value": "4px",
      "type": "spacing"
    },
    "100": {
      "value": "8px",
      "type": "spacing"
    },
    "150": {
      "value": "12px",
      "type": "spacing"
    },
    "200": {
      "value": "16px",
      "type": "spacing"
    },
    "250": {
      "value": "20px",
      "type": "spacing"
    },
    "300": {
      "value": "24px",
      "type": "spacing"
    },
    "350": {
      "value": "28px",
      "type": "spacing"
    },
    "400": {
      "value": "32px",
      "type": "spacing"
    },
    "450": {
      "value": "36px",
      "type": "spacing"
    },
    "500": {
      "value": "40px",
      "type": "spacing"
    },
    "550": {
      "value": "44px",
      "type": "spacing"
    },
    "600": {
      "value": "48px",
      "type": "spacing"
    },
    "650": {
      "value": "52px",
      "type": "spacing"
    },
    "700": {
      "value": "56px",
      "type": "spacing"
    },
    "750": {
      "value": "60px",
      "type": "spacing"
    },
    "800": {
      "value": "64px",
      "type": "spacing"
    }
  },
  "letter-spacing": {
    "0": {
      "value": "0px",
      "type": "letterSpacing"
    },
    "78": {
      "value": "0.11px",
      "type": "letterSpacing"
    },
    "85": {
      "value": "0.12px",
      "type": "letterSpacing"
    },
    "100": {
      "value": "0.14px",
      "type": "letterSpacing"
    },
    "114": {
      "value": "0.16px",
      "type": "letterSpacing"
    },
    "128": {
      "value": "0.18px",
      "type": "letterSpacing"
    },
    "142": {
      "value": "0.20px",
      "type": "letterSpacing"
    },
    "157": {
      "value": "0.22px",
      "type": "letterSpacing"
    },
    "171": {
      "value": "0.24px",
      "type": "letterSpacing"
    },
    "185": {
      "value": "0.26px",
      "type": "letterSpacing"
    }
  }
}