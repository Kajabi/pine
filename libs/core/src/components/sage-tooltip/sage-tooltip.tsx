import { Component, Element, Event, Host, Prop, State, h, EventEmitter, Method, Watch } from '@stencil/core';
import {
  // createObserver,
  positionTooltip
} from '../../utils/overlay';

/**
 * @slot (default) - The tooltip's target element
 * @slot content - HTML content for the tooltip
 *
 * @part arrow - The arrow attached to the tooltip content.
 * @part content - The tooltip content.
 * @part trigger - The tooltip trigger.
 */

@Component({
  tag: 'sage-tooltip',
  styleUrl: 'sage-tooltip.scss',
  shadow: true,
})
export class SageTooltip {
  // private arrowEl: HTMLElement | null;
  // private triggerEl: HTMLElement | null;
  private contentEl: HTMLElement | null;
  // private overlayObserver!: IntersectionObserver;

  /**
   * Reference to the Host element
   */
  @Element() el: HTMLDivElement;

  /**
   * Determines where the tooltip is open
   */
  @State() isOpen = false;

  /**
   * Determines if the overlay is being observed
   */
  // @State() private intersectionObserver: IntersectionObserver;

  /**
   * Content for the tooltip. If HTML is required, use the content slot
   */
  @Prop() content: '';

  /**
   * If populated, this will be the tooltip content's width
   */
  @Prop() customWidth?: number;

  /**
   * Id used to reference the component
   */
  @Prop() componentId: '';

  /**
   * Determines whether or not the tooltip have an arrow
   */
  @Prop() hasArrow?: boolean;


  /**
   * Enable this option when using the content slot
   */
  @Prop() htmlContent?: boolean;

  /**
   * Determines the preferred position of the tooltip
   */
  @Prop() placement:
    'top'
    | 'top-start'
    | 'top-end'
    | 'right'
    | 'right-start'
    | 'right-end'
    | 'bottom'
    | 'bottom-start'
    | 'bottom-end'
    | 'left'
    | 'left-start'
    | 'left-end';

  /**
   * Determines whether or not the tooltip is visible
   */
  @Prop({mutable: true, reflect: true}) opened = false;

  // eslint-disable-next-line @stencil/no-unused-watch
  @Watch('opened')
  handleOpenToggle() {
    console.log('before isOpened', this.opened);

    if(this.opened) {
      this.handleShow();
      this.sageTooltipShow.emit();
      const testing = this.sageTooltipShow.emit();
      console.log('testing: ', testing);
    } else {
      this.handleHide();
      this.sageTooltipHide.emit();
    }
    console.log('after isOpened', this.opened);
  }

  /**
   * Determines whether or not the tooltip have an arrow
   */
  @Prop() disabled = false;

  // TODO Q update this state prop
  // eslint-disable-next-line @stencil/no-unused-watch
  @Watch('disabled')
  handleDisabled() {
    console.log('disabled')
  }

  // TODO Q: make better events, maybe before and after hide and show
  /**
   * Emitted after a tooltip is closed
   */
  @Event() sageTooltipHide: EventEmitter;

  /**
   * Emitted after a tooltip is shown
   */
  @Event() sageTooltipShow: EventEmitter;

  // connectedCallback() {}
  // TODO Q: not working as expected, revisit before putting in review
  // connectedCallback(): void {
  //   console.log('connected callback enter: ', this.intersectionObserver );
  //   createObserver(
  //     this.intersectionObserver,
  //     '0px',
  //     this.contentEl,
  //     () => {
  //       // this.opened = true;
  //       console.log('connected callback');
  //     }
  //   );
  // }

  disconnectedCallback() {
    this.el.removeEventListener('focus', this.handleFocus);
  }

  componentWillLoad() {
    console.log('opened: ', this.opened);
    if (this.opened) {
      this.showTooltip();
    }
  }

  // componentDidLoad() {}

  // componentWillRender() {}

  // componentShouldUpdate(newVal: any, oldVal: any, propName: string) {}

  // componentWillUpdate() {}

  componentDidUpdate() {
    if (this.opened) {
      this.showTooltip();
    }
  }

  componentDidRender() {
    console.log('didRender');
    positionTooltip(this.el, this.placement, this.contentEl);
  }

  // Note to self: Should be click by `mouseover` is causing too many repaints. May need to debounce?
  // @Listen('click', { capture: true })
  // handleClick() {
  //   const tooltipContent = this.el.shadowRoot.querySelector('sage-tooltip__content');
  //   // this.isOpen = !this.isOpen;
  //   tooltipContent.classList.add('is-open');
  // }

  @Method()
  async showTooltip() {
    console.log('showTooltip');

    if(!this.disabled) {
      this.opened = true;
      // TODO: need to use block / none but the tooltip content width and height are needed for calculations
      // this.contentEl.style.display = 'block';
      if(this.contentEl) {
        this.contentEl.style.opacity = '1';
        this.contentEl.style.visibility = 'visible';
      }
    }
  }

  @Method()
  async hideTooltip() {
    this.opened = false;
    // TODO: need to use block / none but the tooltip content width and height are needed for calculations
    // this.contentEl.style.display = '';
    this.contentEl.style.opacity = '0';
    this.contentEl.style.visibility = 'hidden';
  }

  private handleShow = () => {
    // Do not show the overlay if disabled
    if(this.disabled) {
      console.log('inside disabled');
      return
    }
    console.log('disabled');
    this.showTooltip();
  };

  private handleHide = () => {
    this.hideTooltip();
  };

  private handleFocus = () => {
    // TODO Q: add focus functionality
    console.log('handle focus');
    this.showTooltip();
  };

  private handleBlur = () => {
    // TODO Q: add blur functionality
    this.showTooltip();
  };

  render() {
    return (
      <Host
        class={{'sage-tooltip--has-html-content': this.htmlContent}}
        disabled={this.disabled}
        hasArrow={this.hasArrow}
        onMouseEnter={this.handleShow}
        onMouseLeave={this.handleHide}
        onFocusin={this.handleFocus}
        onFocusout={this.handleBlur}
        // onClick={this.handleClick}
      >
        <div
          class={`
            sage-tooltip
            ${this.placement ? (`sage-tooltip--${this.placement}`) : ''}
            ${this.opened ? 'is-open' : ''}

          `}
        >
          <span
            aria-describedby={this.componentId}
            part="trigger"
            class="sage-tooltip__trigger"
            // ref={(el) => (this.triggerEl = el)}
          >
            <slot />
          </span>

          <div class="sage-tooltip__content"
            aria-hidden={this.opened ? 'false' : 'true'}
            aria-live={this.opened ? 'polite' : 'off'}
            id={this.componentId}
            part="content"
            ref={(el) => (this.contentEl = el)}
            role="tooltip"
          >
            <slot
              name="content"
            ></slot>
            {this.content}
            {this.hasArrow && (
              <div
                class="sage-tooltip__arrow"
                aria-hidden="true"
                part="arrow"
                // ref={(el) => (this.arrowEl = el)}
              ></div>
            )}
          </div>
        </div>
      </Host>
    );
  }
}
