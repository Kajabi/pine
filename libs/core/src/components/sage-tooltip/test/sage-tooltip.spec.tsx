import { newSpecPage } from '@stencil/core/testing';
import { SageTooltip } from '../sage-tooltip';

describe('sage-tooltip', () => {
  // it('is rendered', async () => {
  //   const { root } = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `<sage-tooltip></sage-tooltip>`
  //   });
  //   expect(root).toEqualHtml(`
  //   <sage-tooltip>
  //     <mock:shadow-root>
  //     <div class="sage-tooltip>
  //       <span class="sage-tooltip__trigger" part="trigger">
  //         <slot></slot>
  //       </span>
  //       <div class="sage-tooltip__content" aria-hidden="true" aria-live="off" part="content" role="tooltip">
  //         <slot name="content"></slot>
  //       </div>
  //     </div>
  //     </mock:shadow-root>
  //   </sage-tooltip>
  //   `);
  // });

  // it('renders the trigger', async () => {
  //   const { root } = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `
  //     <sage-tooltip>
  //       <a href="#">Trigger</a>
  //     </sage-tooltip>`
  //   });
  //   expect(root).toEqualHtml(`
  //   <sage-tooltip>
  //     <mock:shadow-root>
  //     <div class="sage-tooltip">
  //       <span class="sage-tooltip__trigger" part="trigger">
  //         <slot></slot>
  //       </span>
  //       <div class="sage-tooltip__content" aria-hidden="true" aria-live="off" part="content" role="tooltip">
  //         <slot name="content"></slot>
  //       </div>
  //     </div>
  //     </mock:shadow-root>
  //     <a href="#">Trigger</a>
  //   </sage-tooltip>
  //   `);
  // });

  // it('renders the slot content', async () => {
  //   const { root } = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `
  //     <sage-tooltip>
  //       <p slot="content">this is the slotted tooltip content</p>
  //     </sage-tooltip>`
  //   });
  //   expect(root).toEqualHtml(`
  //   <sage-tooltip>
  //     <mock:shadow-root>
  //     <div class="sage-tooltip">
  //       <span class="sage-tooltip__trigger" part="trigger">
  //         <slot></slot>
  //       </span>
  //       <div class="sage-tooltip__content" aria-hidden="true" aria-live="off" part="content" role="tooltip">
  //         <slot name="content"></slot>
  //       </div>
  //     </div>
  //     </mock:shadow-root>
  //     <p slot="content">this is the slotted tooltip content</p>
  //   </sage-tooltip>
  //   `);
  // });

  // it('renders the placement', async () => {
  //   const { root } = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `<sage-tooltip placement="right-start"></sage-tooltip>`
  //   });

  //   const placement = root.shadowRoot.querySelector('.sage-tooltip--right-start');
  //   expect(placement).not.toBeNull();
  // });

  it('renders with disabled', async () => {
    // arrange
    // act
    // assert
  })

  // it('renders with disabled', async () => {
  //   const page = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `
  //     <sage-tooltip>
  //       <a href="#">Trigger</a>
  //     </sage-tooltip>`
  //   })

  //   page.root.disabled = ''

  //   const component = page.root;
  //   const defaultSlot = component.shadowRoot.querySelector<HTMLSlotElement>(
  //     '.sage-tooltip__trigger slot'
  //   );

  //   defaultSlot.assignedNodes = () =>
  //     component.querySelectorAll('> *') as unknown as Node[]
  //   expect(page.root).toMatchSnapshot()
  // })

  it('showToolip', async() => {
    const page = await newSpecPage({
      components: [SageTooltip],
      html: `
      <sage-tooltip>
        <a href="#">Trigger</a>
      </sage-tooltip>`
    });

    // const foo = page.body.querySelector('foo-component');
    const component = await page.root?.find('sage-tooltip');

    if (!component) {
        throw new Error('Could not find Foo component');
    }

    await expect(component.showTooltip()).toBeTruthy();
  });

  // it('should call method to show', async () => {
  //   const page = await newSpecPage({
  //     components: [SageTooltip],
  //     html: `<sage-tooltip placement="right-start"></sage-tooltip>`
  //   });

  //   const component = await page.root.find('sage-tooltip');

  //   await page.root.showTooltip();
  //   await page.waitForChanges();

  //   // tried in the spec just to be sure
  //   // const wrapper = await page.root.shadowRoot.querySelector('.sage-tooltip__content').isVisible;
  //   // expect(wrapper).
  // });
});
