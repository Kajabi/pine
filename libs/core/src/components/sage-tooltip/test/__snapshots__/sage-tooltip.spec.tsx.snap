// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sage-tooltip renders with arrow 1`] = `
<sage-tooltip has-arrow="true" hasarrow="">
  <mock:shadow-root>
    <div class="sage-tooltip">
      <span class="sage-tooltip__trigger" part="trigger">
        <slot></slot>
      </span>
      <div aria-hidden="true" aria-live="off" class="sage-tooltip__content" part="content" role="tooltip">
        <slot name="content"></slot>
        <div aria-hidden="true" class="sage-tooltip__arrow" part="arrow"></div>
      </div>
    </div>
  </mock:shadow-root>
</sage-tooltip>
`;

exports[`sage-tooltip renders with disabled 1`] = `
<sage-tooltip>
  <mock:shadow-root>
    <div class="sage-tooltip">
      <span class="sage-tooltip__trigger" part="trigger">
        <slot></slot>
      </span>
      <div aria-hidden="true" aria-live="off" class="sage-tooltip__content" part="content" role="tooltip">
        <slot name="content"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <a href="#">
    Trigger
  </a>
</sage-tooltip>
`;
